connectors/sink-google-sheets/internal/sink.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Linkall Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+	"strconv"
+	ce "github.com/cloudevents/sdk-go/v2"
+	cdkgo "github.com/linkall-labs/cdk-go"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+	"google.golang.org/api/sheets/v4"
+)
+
+var _ cdkgo.Sink = &GoogleSheetSink{}
+
+func NewGoogleSheetSink() cdkgo.Sink {
+	return &GoogleSheetSink{}
+}
+
+type GoogleSheetSink struct {
+	config *GoogleSheetConfig
+}
+
+func (s *GoogleSheetSink) Initialize(ctx context.Context, cfg cdkgo.ConfigAccessor) error {
+	// TODO
+	s.config = cfg.(*GoogleSheetConfig)
+	
+	return nil
+}
+
+func (s *GoogleSheetSink) Name() string {
+	// TODO
+	return "GoogleSheetSink"
+}
+
+func (s *GoogleSheetSink) Destroy() error {
+	// TODO
+	return nil
+}
+
+func (s *GoogleSheetSink) Arrived(ctx context.Context, events ...*ce.Event) cdkgo.Result {
+	// TODO
+	for _, event := range events {
+
+		s.saveDataToSpreadsheet(event)
+	}
+	return cdkgo.SuccessResult
+}
+
+
+
+func (s *GoogleSheetSink) saveDataToSpreadsheet(event *ce.Event) {
+
+		//Create API Context
+	ctx := context.Background()
+
+	// authenticate and get configuration
+	config, err := google.JWTConfigFromJSON([]byte(s.config.Credentials), "https://www.googleapis.com/auth/spreadsheets")
+		if err != nil {
+			log.Fatalf("Failed to authenticate google service accounts key %v", err)
+			return
+		}
+
+	//Create Client
+	client := config.Client(ctx)
```

Comment:
It doesn't need to new a client for every event arrived. 
you can put it in function Initialize

Reply:
I don't understand this

Reply:
Done

Reply:
it is unnecessary for new a client for every event that arrived. you can add a property `client *sheets.Service`  in struct`GoogleSheetSink. Then put below code in function Initialize

```go
  config, err := google.JWTConfigFromJSON([]byte(s.config.Credentials), "https://www.googleapis.com/auth/spreadsheets")
      if err != nil {
	      return err
      }

	//Create Client
	client := config.Client(context.Background())

	//Create Service using Client
	srv, err := sheets.NewService(context.Background(), option.WithHTTPClient(client))
	if err != nil {
        return err
	}
        s.client = srv
```